File: ChessServer/src/main/java/org/example/logicCode/GameState.java
package org.example.logicCode;
import java.util.List;
import java.util.ArrayList;
public class GameState {
    private char[][] board;
    private boolean whiteToMove;
    private List<Move> moveHistory;

    public GameState() {
        board = new char[8][8];
        whiteToMove = true;
        moveHistory = new ArrayList<>();
        initializeBoard();
    }

    private void initializeBoard() {
        // Initialize the board with starting positions
        String[] startingPositions = {
                "rnbqkbnr",
                "pppppppp",
                "--------",
                "--------",
                "--------",
                "--------",
                "PPPPPPPP",
                "RNBQKBNR"
        };
        for (int i = 0; i < 8; i++) {
            board[i] = startingPositions[i].toCharArray();
        }
    }

    public char[][] getBoard() {
        return board;
    }

    public boolean isWhiteToMove() {
        return whiteToMove;
    }

    public List<Move> getValidMoves() {
        // Generate and return a list of valid moves
        List<Move> moves = new ArrayList<>();
        for (int r = 0; r < 8; r++) {
            for (int c = 0; c < 8; c++) {
                char piece = board[r][c];
                if (piece != '-' && Character.isUpperCase(piece) == whiteToMove) {
                    // Add move generation logic here
                    // Example: moves.addAll(getPieceMoves(r, c, piece));
                }
            }
        }
        return moves;
    }

    public void makeMove(Move move) {
        // Execute a move
        board[move.getEndRow()][move.getEndCol()] = board[move.getStartRow()][move.getStartCol()];
        board[move.getStartRow()][move.getStartCol()] = '-';
        whiteToMove = !whiteToMove;
        moveHistory.add(move);
    }

    public void undoMove() {
        if (!moveHistory.isEmpty()) {
            Move move = moveHistory.remove(moveHistory.size() - 1);
            board[move.getStartRow()][move.getStartCol()] = board[move.getEndRow()][move.getEndCol()];
            board[move.getEndRow()][move.getEndCol()] = move.getCapturedPiece();
            whiteToMove = !whiteToMove;
        }
    }

    public boolean isCheckmate() {
        // Check for checkmate
        return false;
    }

    public boolean isStalemate() {
        // Check for stalemate
        return false;
    }

}

File: MyApplication2/app/src/main/java/com/murach/myapplication/newLogicBot/GameState.kt
package com.murach.myapplication.newLogicBot

class GameState {
    val board: Array<CharArray> = Array(8) { CharArray(8) }
    var whiteToMove: Boolean = true
    private val moveHistory: MutableList<Move> = mutableListOf()

    init {
        initializeBoard()
    }

    private fun initializeBoard() {
        val startingPositions = arrayOf(
            "rnbqkbnr",
            "pppppppp",
            "--------",
            "--------",
            "--------",
            "--------",
            "PPPPPPPP",
            "RNBQKBNR"
        )
        for (i in startingPositions.indices) {
            board[i] = startingPositions[i].toCharArray()
        }
    }

    fun getValidMoves(): List<Move> {
        val validMoves = mutableListOf<Move>()
        for (r in board.indices) {
            for (c in board[r].indices) {
                val piece = board[r][c]
                if (piece != '-' && isPieceWhite(piece) == whiteToMove) {
                    validMoves.addAll(getValidMovesForPiece(r, c))
                }
            }
        }
        return validMoves
    }

    fun getValidMovesForPiece(row: Int, col: Int): List<Move> {
        val piece = board[row][col]
        val moves = mutableListOf<Move>()
        if (piece == 'P' || piece == 'p') {
            val direction = if (piece == 'P') -1 else 1
            val startRow = if (piece == 'P') 6 else 1
            if (board[row + direction][col] == '-') {
                moves.add(Move(row, col, row + direction, col, piece, '-'))
                if (row == startRow && board[row + 2 * direction][col] == '-') {
                    moves.add(Move(row, col, row + 2 * direction, col, piece, '-'))
                }
            }
            if (col - 1 >= 0 && board[row + direction][col - 1] != '-' && isPieceWhite(board[row + direction][col - 1]) != whiteToMove) {
                moves.add(Move(row, col, row + direction, col - 1, piece, board[row + direction][col - 1]))
            }
            if (col + 1 < 8 && board[row + direction][col + 1] != '-' && isPieceWhite(board[row + direction][col + 1]) != whiteToMove) {
                moves.add(Move(row, col, row + direction, col + 1, piece, board[row + direction][col + 1]))
            }
        }
        // Add logic for other pieces
        return moves
    }

    private fun isPieceWhite(piece: Char): Boolean {
        return piece.isUpperCase()
    }

    fun makeMove(move: Move) {
        board[move.endRow][move.endCol] = board[move.startRow][move.startCol]
        board[move.startRow][move.startCol] = '-'
        whiteToMove = !whiteToMove
        moveHistory.add(move)
    }

    fun undoMove() {
        if (moveHistory.isNotEmpty()) {
            val move = moveHistory.removeAt(moveHistory.size - 1)
            board[move.startRow][move.startCol] = board[move.endRow][move.endCol]
            board[move.endRow][move.endCol] = move.capturedPiece
            whiteToMove = !whiteToMove
        }
    }

    fun isCheckmate(): Boolean {
        // Check for checkmate
        return false
    }

    fun isStalemate(): Boolean {
        // Check for stalemate
        return false
    }
}
